packages/grid/x-data-grid/src/hooks/features/rows/useGridRows.ts

```diff
@@ -325,6 +326,63 @@ export const useGridRows = (
     [apiRef, logger],
   );
 
+  const replaceRows = React.useCallback<GridRowApi['replaceRows']>(
+    (firstRowToRender, lastRowToRender, newRows) => {
+      if (props.signature === GridSignature.DataGrid) {
+        throw new Error(
+          [
+            "MUI: You can't replace rows using `apiRef.current.replaceRows` on the DataGrid.",
+            'You need to upgrade to the DataGridPro component to unlock this feature.',
+          ].join('\n'),
+        );
+      }
+
+      const newRowsIds = newRows.map((row) => row.id);
+      const allRows = gridRowIdsSelector(apiRef);
+      const updatedRows = [...allRows];
+      updatedRows.splice(firstRowToRender, lastRowToRender - firstRowToRender, ...newRowsIds);
+
+      const idRowsLookup = gridRowsLookupSelector(apiRef);
+      const idToIdLookup = gridRowsIdToIdLookupSelector(apiRef);
+      const tree = gridRowTreeSelector(apiRef);
+      const updatedIdRowsLookup = { ...idRowsLookup };
+      const updatedIdToIdLookup = { ...idToIdLookup };
+      const updatedTree = { ...tree };
+      const rowIdsToBeReplaced = allRows.slice(firstRowToRender, lastRowToRender);
+      rowIdsToBeReplaced.forEach((id) => {
+        delete updatedIdRowsLookup[id];
+        delete updatedIdToIdLookup[id];
+        delete updatedTree[id];
+      });
+
+      const newIdRowsLookup = newRows.reduce(
```

Comment:
We should avoid looping several time the list of the rows

Reply:
The size of the `newRows` array will never be too big (depends on the viewport size and the buffer). Either way, I updated it from 3 loops to 1. It looks more condensed that way.